lib/stages/HFBAccumulate.cpp

```diff
@@ -219,17 +212,49 @@ void HFBAccumulate::main_thread() {
         DEBUG(
             "fpga_seq_start: {:d}, fpga_seq_num_end: {:d}, num_frames * num_samples: {:d}, fpga % "
             "(align): {:d}",
-            get_fpga_seq_start_hfb(in_buf, in_buffer_ID), fpga_seq_num_end,
+            in_frame.fpga_seq_start, fpga_seq_num_end,
             _num_frames_to_integrate * _samples_per_data_set,
-            get_fpga_seq_start_hfb(in_buf, in_buffer_ID)
+            in_frame.fpga_seq_start
                 % (_num_frames_to_integrate * _samples_per_data_set));
 
         // Get the no. of lost samples in this frame
         total_lost_timesamples +=
-            get_lost_timesamples(compressed_lost_samples_buf, compress_buffer_ID);
+            get_lost_timesamples(cls_buf, cls_frame_id);
+
+        int32_t lost_in_frame = get_lost_timesamples(cls_buf, cls_frame_id);
+        
+        total_lost_timesamples += lost_in_frame;
+
+        int32_t samples_in_frame = _samples_per_data_set - lost_in_frame;
+
+        // We are calculating the weights by differencing even and odd samples.
+        // Every even sample we save the set of visibilities...
+        if (frame_count % 2 == 0) {
+            std::memcpy(hfb_even.data(), input, _num_frb_total_beams * _factor_upchan);
+            samples_even = samples_in_frame;
+        }
+        // ... every odd sample we accumulate the squared differences into the weight dataset
+        // NOTE: this incrementally calculates the variance, but eventually
+        // output_frame.weight will hold the *inverse* variance
+        // TODO: we might need to account for packet loss in here too, but it
+        // would require some awkward rescalings
+        else {
+            for (size_t i = 0; i < _num_frb_total_beams * _factor_upchan; i++) {
+                // NOTE: avoid using the slow std::complex routines in here
+                //INFO("hfb_even[{}]: {}", i, hfb_even[i]);
+                float d = input[i] - hfb_even[i];
```

Comment:
If I had to pick one place that I thought was responsible for the incorrect weights, it is here. Both the array `input` and vector `hfb_even` are `int32_t` typed even though the underlying data is floating point. This subtraction operation will do an integer subtraction, which is then converted into a float. That will give the wrong answer, you need to do a floating point subtraction initially.

Ultimately the issue here is that none of the data is `int32_t`, if you type all the arrays and vectors appropriately this issue should disappear.